esp/bindings/http/client/httpclient.cpp

```diff
@@ -405,13 +405,134 @@ int CHttpClient::sendRequest(const char* method, const char* contenttype, String
+            switch (*name)
+            {
+            case 'H':
+            case 'h':
+                if (strieq(name, "Host"))
+                    continue;
+                break;
+            case 'C':
+            case 'c':
+                if (strnicmp(name, "Content-", 8)==0)
+                {
+                    if (strieq(name+8, "Length") || strieq(name+8, "Type"))
+                        continue;
+                }
+                break;
+            default:
+                break;
+            }
+            copyTo.setHeader(name, pair.item(1));
+        }
+    }
+}
+
+void copyCookies(CHttpMessage &copyTo, CHttpMessage &copyFrom, const char *host)
 {
+    IArrayOf<CEspCookie>& cookies = copyFrom.queryCookies();
+    ForEachItemIn(x, cookies)
+    {
+        CEspCookie* cookie = &cookies.item(x);
+        if(!cookie)
+            continue;
+        cookie->setHost(host); //unfortunately changes copyFrom cookie... should make a true copy of cookie
+        copyTo.addCookie(cookie);
+    }
+}
+
+int CHttpClient::proxyRequest(IHttpMessage *request, IHttpMessage *response)
+{
+    CHttpRequest *forwardRequest = static_cast<CHttpRequest*>(request);
+    assertex(forwardRequest != nullptr);
+
+    CHttpResponse *forwardResponse = static_cast<CHttpResponse*>(response);
+    assertex(forwardResponse != nullptr);
+
     StringBuffer errmsg;
     if(connect(errmsg) < 0)
     {
-        response.append(errmsg);
+        forwardResponse->setContent(errmsg);
+        forwardResponse->setContentType(HTTP_TYPE_TEXT_PLAIN);
+        return -1;
+    }
+
+    Owned<CHttpRequest> httprequest = new CHttpRequest(*m_socket);
+    Owned<CHttpResponse> httpresponse = new CHttpResponse(*m_socket);
+
+    httprequest->setMethod(forwardRequest->queryMethod());
+    httprequest->setVersion("HTTP/1.1");
+
+    if(m_proxy.length() <= 0)
+        httprequest->setPath(m_path.get());
+    else
+        httprequest->setPath(m_url.get());
+
+    httprequest->setHost(m_host.get());
+    httprequest->setPort(m_port);
+
+    copyHeaders(*httprequest, *forwardRequest);
+
+    StringBuffer contentType;
+    forwardRequest->getContentType(contentType);
+
+    httprequest->setContentType(contentType);
+    httprequest->setContent(forwardRequest->queryContent());
+
+    if (getEspLogLevel()>LogNormal)
+    {
+        StringBuffer s;
+        DBGLOG("Content type: %s", forwardRequest->getContentType(s).str());
+        DBGLOG("Request content: %s", forwardRequest->queryContent());
+    }
+
+    Owned<IMultiException> me = MakeMultiException();
+
+    copyCookies(*httprequest, *forwardRequest, m_host);
+
+    httprequest->send();
+
+    if (m_readTimeoutSecs)
+        httpresponse->setTimeOut(m_readTimeoutSecs);
+    httpresponse->receive(false, me);  // MORE - pass in IMultiException if we want to see exceptions (which are not fatal)
```

Comment:
Do we need to 'forward' the me into forwardResponse?

Reply:
Good point. Hmmm I'll have to look into what these represent.  At least make sure they are logged if not forwarded.